### DISCUSSION TITLE
About TGFX roadmap

### DISCUSSION THREAD

**Discussion by xdestiny110** [Tencent/tgfx#63] [General] [answered]
This is a great repository and I'm very much looking forward to its development. I'm wondering what your plans are for the future of this repository.
BTW, compared to Skia, TGFX has removed many internal renderers. How did you evaluate the impact on performance from removing those renderers?

**Comment by domchen** [comment, marked as answer]
If I can find the time, I may write an article that explores the core architecture of tgfx and the different design choices that set tgfx apart from Skia.
When referring to 'renderers', I believe you are specifically referring to path renderers. Indeed, tgfx only implements path renderers of triangulation and software rasterization. We conducted a test by disabling all GPU path renderers in Skia. To our surprise, we discovered that there was no decline in performance. In fact, we even observed a slight performance improvement. This result is not that hard to accept, especially considering that CoreGraphics on Apple platforms has never implemented any direct GPU path renderers throughout these years. Nonetheless, the vector-based graphics continue to exhibit superior smoothness on iOS and macOS.  Apple utilizes hardware buffers to efficiently rasterize paths into textures without the need for an uploading process. This approach is the actual key for enhancing path rendering performance, rather than implementing various GPU path renderers.

**Reply by xdestiny110** [reply]
How should I understand the meaning of the sentence `Apple utilizes hardware buffers to efficiently rasterize paths into textures without the need for an uploading process`? Does it refer to something like CVPixelBuffer?

**Reply by domchen** [reply]
Yes, It's CVPixelBuffer on apple platforms. The corresponding class on Android is HardwareBuffer, but it is only available since Android 8.0. The internal implementation of hardware buffers inside the Android system is GraphicBuffer. They are all for the same purpose: allowing sharing memory across various hardware units.
